src/librustc_typeck/check/generator_interior.rs

```diff
@@ -89,15 +107,27 @@ pub fn resolve_interior<'a, 'gcx, 'tcx>(fcx: &'a FnCtxt<'a, 'gcx, 'tcx>,
     // Extract type components
     let types: Vec<_> = types.into_iter().map(|t| t.0).collect();
 
-    let tuple = fcx.tcx.intern_tup(&types, false);
+    let type_list = fcx.tcx.intern_type_list(&types);
+
+    debug!("Types in generator {:?}, span = {:?}", type_list, body.value.span);
 
-    debug!("Types in generator {:?}, span = {:?}", tuple, body.value.span);
+    let mut counter = 0;
+    let type_list = fcx.tcx.fold_regions(&type_list, &mut false, |_, current_depth| {
```

Comment:
Maybe add a comment to the following effect:

The types in the generator interior contain lifetimes local to the generator itself, which should not be exposed outside of the generator. Therefore, we replace these lifetimes with existentially-bound lifetimes, which reflect the exact value of the lifetimes not being known by users.

These lifetimes are used in auto trait impl checking (for example, if a `Sync` generator contains an `&'α T`, we need to check whether `&'α T: Sync`), so knowledge of the exact relationships between them isn't particularly important.

Reply:
> Note that lifetimes in the type of a local that is currently dead need not be alive. It's up to the generator to ensure that captured lifetimes that are actually used are alive.

I'm not sure what you mean by this. I added your other paragraphs as comments.